File: scala/src/main/scala/datastructures/BinarySearchTree.scala
package datastructures

/**
 * Binary search tree with standard operations.
 * Time: O(log n) average, O(n) worst case
 * Space: O(n)
 */
class BinarySearchTree[T](using ord: Ordering[T]):
  private case class Node(var value: T, var left: Node = null, var right: Node = null)

  private var root: Node = null
  private var count = 0

  def insert(value: T): this.type =
    root = insertNode(root, value)
    this

  private def insertNode(node: Node, value: T): Node =
    if node == null then
      count += 1
      Node(value)
    else if ord.lt(value, node.value) then
      node.left = insertNode(node.left, value)
      node
    else if ord.gt(value, node.value) then
      node.right = insertNode(node.right, value)
      node
    else
      node.value = value
      node

  def contains(value: T): Boolean = findNode(root, value) != null

  private def findNode(node: Node, value: T): Node =
    if node == null then null
    else if ord.lt(value, node.value) then findNode(node.left, value)
    else if ord.gt(value, node.value) then findNode(node.right, value)
    else node

  def remove(value: T): Boolean =
    val originalCount = count
    root = removeNode(root, value)
    count < originalCount

  private def removeNode(node: Node, value: T): Node =
    if node == null then null
    else if ord.lt(value, node.value) then
      node.left = removeNode(node.left, value)
      node
    else if ord.gt(value, node.value) then
      node.right = removeNode(node.right, value)
      node
    else
      count -= 1
      if node.left == null then node.right
      else if node.right == null then node.left
      else
        val successor = findMin(node.right)
        node.value = successor.value
        count += 1
        node.right = removeNode(node.right, successor.value)
        node

  private def findMin(node: Node): Node =
    if node.left == null then node
    else findMin(node.left)

  def min: Option[T] =
    if root == null then None
    else Some(findMin(root).value)

  def max: Option[T] =
    if root == null then None
    else
      var current = root
      while current.right != null do current = current.right
      Some(current.value)

  def isEmpty: Boolean = count == 0

  def size: Int = count

  def clear(): this.type =
    root = null
    count = 0
    this

  def inorder: List[T] =
    val builder = List.newBuilder[T]
    def traverse(node: Node): Unit =
      if node != null then
        traverse(node.left)
        builder += node.value
        traverse(node.right)
    traverse(root)
    builder.result()

  def preorder: List[T] =
    val builder = List.newBuilder[T]
    def traverse(node: Node): Unit =
      if node != null then
        builder += node.value
        traverse(node.left)
        traverse(node.right)
    traverse(root)
    builder.result()

  def postorder: List[T] =
    val builder = List.newBuilder[T]
    def traverse(node: Node): Unit =
      if node != null then
        traverse(node.left)
        traverse(node.right)
        builder += node.value
    traverse(root)
    builder.result()

  def levelOrder: List[T] =
    if root == null then List.empty
    else
      val builder = List.newBuilder[T]
      val queue = scala.collection.mutable.Queue[Node](root)
      while queue.nonEmpty do
        val node = queue.dequeue()
        builder += node.value
        if node.left != null then queue.enqueue(node.left)
        if node.right != null then queue.enqueue(node.right)
      builder.result()

File: haskell/src/DataStructures/BinarySearchTree.hs
{-|
Module      : DataStructures.BinarySearchTree
Description : Binary search tree implementation
Stability   : stable

A purely functional binary search tree.
Time: O(log n) average, O(n) worst case
Space: O(n)
-}
module DataStructures.BinarySearchTree
  ( BST
  , empty
  , isEmpty
  , insert
  , contains
  , remove
  , findMin
  , findMax
  , size
  , inorder
  , preorder
  , postorder
  , levelOrder
  , fromList
  , toList
  ) where

-- | A binary search tree
data BST a = Empty | Node a (BST a) (BST a)
  deriving (Show, Eq)

-- | Create an empty tree
empty :: BST a
empty = Empty

-- | Check if tree is empty
isEmpty :: BST a -> Bool
isEmpty Empty = True
isEmpty _     = False

-- | Insert a value into the tree
insert :: Ord a => a -> BST a -> BST a
insert x Empty = Node x Empty Empty
insert x (Node v left right)
  | x < v     = Node v (insert x left) right
  | x > v     = Node v left (insert x right)
  | otherwise = Node x left right  -- Replace existing value

-- | Check if value exists in tree
contains :: Ord a => a -> BST a -> Bool
contains _ Empty = False
contains x (Node v left right)
  | x < v     = contains x left
  | x > v     = contains x right
  | otherwise = True

-- | Remove a value from the tree
remove :: Ord a => a -> BST a -> BST a
remove _ Empty = Empty
remove x (Node v left right)
  | x < v     = Node v (remove x left) right
  | x > v     = Node v left (remove x right)
  | otherwise = removeNode (Node v left right)

-- | Helper to remove the current node
removeNode :: Ord a => BST a -> BST a
removeNode (Node _ Empty right) = right
removeNode (Node _ left Empty)  = left
removeNode (Node _ left right)  =
  let minVal = getMin right
  in Node minVal left (remove minVal right)
removeNode Empty = Empty

-- | Get minimum value (helper)
getMin :: BST a -> a
getMin (Node v Empty _) = v
getMin (Node _ left _)  = getMin left
getMin Empty            = error "Empty tree has no minimum"

-- | Find minimum value
findMin :: BST a -> Maybe a
findMin Empty           = Nothing
findMin (Node v Empty _) = Just v
findMin (Node _ left _)  = findMin left

-- | Find maximum value
findMax :: BST a -> Maybe a
findMax Empty            = Nothing
findMax (Node v _ Empty) = Just v
findMax (Node _ _ right) = findMax right

-- | Get number of nodes
size :: BST a -> Int
size Empty             = 0
size (Node _ left right) = 1 + size left + size right

-- | Inorder traversal (sorted order)
inorder :: BST a -> [a]
inorder Empty             = []
inorder (Node v left right) = inorder left ++ [v] ++ inorder right

-- | Preorder traversal
preorder :: BST a -> [a]
preorder Empty             = []
preorder (Node v left right) = v : preorder left ++ preorder right

-- | Postorder traversal
postorder :: BST a -> [a]
postorder Empty             = []
postorder (Node v left right) = postorder left ++ postorder right ++ [v]

-- | Level order traversal (breadth-first)
levelOrder :: BST a -> [a]
levelOrder tree = go [tree]
  where
    go [] = []
    go (Empty : rest) = go rest
    go (Node v left right : rest) = v : go (rest ++ [left, right])

-- | Create tree from list
fromList :: Ord a => [a] -> BST a
fromList = foldr insert empty . reverse

-- | Convert to sorted list (same as inorder)
toList :: BST a -> [a]
toList = inorder
